### PR TITLE
Clews 14263 part2 analogous years

### PR DESCRIPTION
**Pull request by joysaikat**

### CODE REVIEW

File: api/client/samples/analogous_years/lib/final_ranks_computation.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+import os
+
+import matplotlib
+matplotlib.use('agg')
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+from scipy.special import comb
+
+from . import \
```
Comment (tolothy) [inline api/client/samples/analogous_years/lib/final_ranks_computation.py]: let's try absolute path import

File: api/client/samples/analogous_years/lib/final_ranks_computation_test.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import pandas as pd
+from pandas.util.testing import assert_frame_equal
+
+from .final_ranks_computation import combined_methods_distances
```
Comment (tolothy) [inline api/client/samples/analogous_years/lib/final_ranks_computation_test.py]: absolute import

File: api/client/samples/analogous_years/lib/get_transform_data_test.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
+from api.client.gro_client import GroClient
+import mock
+import pandas as pd
+from pandas.util.testing import assert_frame_equal
+import pytest
+
+from . import get_transform_data
```
Comment (tolothy) [inline api/client/samples/analogous_years/lib/get_transform_data_test.py]: absolute import

Comment (joysaikat) [reply api/client/samples/analogous_years/lib/get_transform_data_test.py]: changed

File: api/client/samples/analogous_years/lib/final_ranks_computation_test.py (hunk 2 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/lib/final_ranks_computation_test.py]: changed

File: api/client/samples/analogous_years/lib/final_ranks_computation.py (hunk 1 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/lib/final_ranks_computation.py]: done

File: api/client/samples/analogous_years/readme.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+given by the time period ± 1 year, ± 2 years, ..., ± N years using `Analogous Years` application.
+For rank computation a user must provide:
+
+1. A time period determined by `initial_date` and `final_date` such that they are within a year of
+each other.
+2. Gro-entity(ies) and corresponding weights to compare the time periods.
+
+## Working Example Using Python
+To get started with `Analogous Years`, users have to install `Gro API Client`
+as detailed [here](https://github.com/gro-intelligence/api-client).
```
Comment (JGrinst) [inline api/client/samples/analogous_years/readme.md]: We can link to our official docs page instead: https://developers.gro-intelligence.com/installation.html

**Review by tolothy — COMMENTED** [review]
could you delete all the __pychache__ dir and all .pyc files?

File: api/client/samples/analogous_years/readme.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
+# Analogous Years
+Users can find similarity based ranks between a time period and a list of time periods 
+given by the time period ± 1 year, ± 2 years, ..., ± N years using `Analogous Years` application.
+For rank computation a user must provide:
```
Comment (JGrinst) [inline api/client/samples/analogous_years/readme.md]: I think the intro should be a bit more accessible.  Perhaps something like:

"The `Analogous Years` application enables users to compare events from a set period of time, to those of the same date range in other years. The application will compute ranks of similarity between the specified period, and the same period from previous or future years. Multiple different inputs can be used in determining the ranks, but to compute these ranks, a user must provide:

File: api/client/samples/analogous_years/readme.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+2. Gro-entity(ies) and corresponding weights to compare the time periods.
+
+## Working Example Using Python
+To get started with `Analogous Years`, users have to install `Gro API Client`
+as detailed [here](https://github.com/gro-intelligence/api-client). 
+
+If the user wants to know which period of time is closest to the time period 
+between 1<sup>st</sup> January 2019 and 31<sup>st</sup> October 2019 with respect to the
+following Gro-entities -
+1. Rainfall (metric_id=2100031, item_id=2039, source_id=35, frequency_id=1)
```
Comment (JGrinst) [inline api/client/samples/analogous_years/readme.md]: We actually have multiple different rainfall sources, so it might be a bit misleading to say this represents rainfall. Source would need to be specified in the name, as well.

My inner Nemo voice also has an aversion to setting ID numbers. But it's probably fine here.

File: api/client/samples/analogous_years/readme.md [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+1. A time period determined by `initial_date` and `final_date` such that they are within a year of
+each other.
+2. Gro-entity(ies) and corresponding weights to compare the time periods.
+
+## Working Example Using Python
+To get started with `Analogous Years`, users have to install `Gro API Client`
+as detailed [here](https://github.com/gro-intelligence/api-client). 
+
+If the user wants to know which period of time is closest to the time period
```
Comment (JGrinst) [inline api/client/samples/analogous_years/readme.md]: which period of time is closest => which period of time is most similar

?

File: api/client/samples/analogous_years/readme.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+print(result)
+```
+### Working Example Using Shell
+To access ranks for the same inputs using command line users may run the following code 
+in the command line from the `/api-client/api/client/samples/analogous_years` directory: 
+```shell script
+analogous_years $ python run_analogous_years.py -m 2100031 2540047 15531082 -i 2039 3457 7382 -r 100000100 -s 35 26 43 -f 1 1 1 --initial_date 2019-01-01 --final_date 2019-10-31
+```
+The ranks will be saved as a csv file in the same directory.
+## Appendix
```
Comment (JGrinst) [inline api/client/samples/analogous_years/readme.md]: Add an extra line here, before appendix

File: api/client/samples/analogous_years/readme.md (hunk 4 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/readme.md]: Thanks, updated!

File: api/client/samples/analogous_years/readme.md (hunk 5 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/readme.md]: Copying your lines, thanks!

File: api/client/samples/analogous_years/readme.md (hunk 7 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/readme.md]: Done, thanks!

File: api/client/samples/analogous_years/readme.md (hunk 8 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/readme.md]: Done

**Review by tolothy — APPROVED** [review]
Look good to me now

File: api/client/samples/analogous_years/readme.md (hunk 6 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/readme.md]: Added the names of the sources associated with each entity.

File: api/client/samples/analogous_years/readme.md [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+as detailed [here](https://developers.gro-intelligence.com/installation.html). 
+
+If the user wants to know which period of time is most similar to the time period 
+between 1<sup>st</sup> January 2019 and 31<sup>st</sup> October 2019 with respect to the
+following Gro-entities -
+1. Rainfall, TRMM (metric_id=2100031, item_id=2039, source_id=35, frequency_id=1)
+2. Land Temperature, MODIS (metric_id=2540047, item_id=3457, source_id=26, frequency_id=1)
+3. Soil moisture, SMOS (metric_id=15531082, item_id=7382, source_id=43, frequency_id=1)
+
+in the US Corn Belt (region_id=100000100) with equal weights given (by default) to
```
Comment (nemozen) [inline api/client/samples/analogous_years/readme.md]: try to link these entities to dictionary https://app.gro-intelligence.com/dictionary/regions/100000100

File: api/client/samples/analogous_years/readme.md [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+# Import Analogous Years
+from api.client.samples.analogous_years import run_analogous_years
+from api.client.samples.analogous_years.lib import final_ranks_computation, 
+    get_transform_data
+
+# Input (Gro Entities)
+
+# Rainfall (modeled) - Precipitation Quantity - 
+# US Corn Belt States (NASA TRMM 3B42RT)
+entity_1 = {'metric_id': 2100031,
```
Comment (nemozen) [inline api/client/samples/analogous_years/readme.md]: data_series or selection

File: api/client/samples/analogous_years/lib/final_ranks_computation.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        combined_items_distances.sort_index(inplace=True)
+    if all_ranks:
+        display_dataframe = combined_items_distances[ranks]
+    else:
+        display_dataframe = combined_items_distances['composite_rank']
+    logger.info("\n Computing analogous years ranks for {} \n".format(file_name))
+    return file_name, display_dataframe
+
+
+def save_to_csv(dataframe_file_name, logger, all_ranks=None, report=True, output_dir=''):
```
Comment (nemozen) [inline api/client/samples/analogous_years/lib/final_ranks_computation.py]: csv and png output should be two different functions

File: api/client/samples/analogous_years/lib/final_ranks_computation.py [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+    start_date = common_start_date(client, entities, provided_start_date)['start_date']
+    if not weights:
+        weights = [1] * len(entities)
+    if enso:
+        entities.append(enso_data(start_date))
+        if enso_weight:
+            weights.append(enso_weight)
+        else:
+            weights.append(1)
+    file_name = get_file_name(client, entities, initial_date, final_date)
```
Comment (nemozen) [inline api/client/samples/analogous_years/lib/final_ranks_computation.py]: this is nicely separable so keep it outside
and make this function only return the dataframe

File: api/client/samples/analogous_years/lib/get_transform_data.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    :param start_date: start-date of the Gro data series
+    :return: A dataframe with an 'end_date' and 'value' column
+    """
+    data = client.get_data_points(**{'metric_id': metric_id,
+                                     'item_id': item_id,
+                                     'region_id': region_id,
+                                     'source_id': source_id,
+                                     'frequency_id': frequency_id,
+                                     'start_date': start_date})
+    data = pd.DataFrame(data)
```
Comment (nemozen) [inline api/client/samples/analogous_years/lib/get_transform_data.py]: use

```
for series in the series you wnat:
   client.add_single_data_series(** series)

client.get_df()
```
this will give you a dataframe that contains the

File: api/client/samples/analogous_years/readme.md [hunk 14]
```diff
@@ -0,0 +1,164 @@
+# Analogous Years
```
Comment (nemozen) [inline api/client/samples/analogous_years/readme.md:1]: let's remove redundant info between readme and notebook now that the latter exists
readme can jus tbe a short cmd line example
and then can point to notebook for more complex usage

**Review by nemozen — APPROVED** [review]
when ready to run the final notebook, you can edit 
api.client.cfg.py and change log level to ERROR

File: api/client/samples/analogous_years/readme.md (hunk 14 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/readme.md:1]: There is still some overlap. However, I have removed much of it.

File: api/client/samples/analogous_years/lib/get_transform_data.py (hunk 13 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/lib/get_transform_data.py]: I have added get_df in the current workflow. 
Note: Assuming series is a dictionary of the form `series =  {'metric_id': metric_id, 'item_id': item_id, 'region_id': region_id, 'frequency_id': frequency_id, 'source_id': source_id, 'start_date': start_date}`, then the code would look like
```
client.add_single_data_series(series)
client.get_df()
```

File: api/client/samples/analogous_years/lib/final_ranks_computation.py (hunk 12 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/lib/final_ranks_computation.py]: Done

File: api/client/samples/analogous_years/lib/final_ranks_computation.py (hunk 11 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/lib/final_ranks_computation.py]: Done

File: api/client/samples/analogous_years/readme.md (hunk 10 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/readme.md]: Done

File: api/client/samples/analogous_years/readme.md (hunk 9 above)
Comment (joysaikat) [reply api/client/samples/analogous_years/readme.md]: Done

File: api/.gitignore [hunk 15]
```diff
@@ -1,2 +1,3 @@
 *.pyc
-*.csv
\ No newline at end of file
+*.csv
+.idea/
```
Comment (nemozen) [inline api/.gitignore]: revert this unrelated change

File: requirements-test.txt [hunk 16]
```diff
@@ -1,5 +1,12 @@
-pytest
-pytest-cov
 dateparser
+dtw==1.3.3
+matplotlib
 mock
-sklearn
```
Comment (nemozen) [inline requirements-test.txt:5]: why change ? https://stackoverflow.com/questions/38733220/difference-between-scikit-learn-and-sklearn

**Review by nemozen — CHANGES_REQUESTED** [review]
couple of very minor comments.

also please remove the warning about  "Panel class is removed from pandas:" in the noteboook... you can either edit it out manually or rerun the notebook with log level set to ERROR in cfg.py

File: requirements-test.txt (hunk 16 above)
Comment (joysaikat) [reply requirements-test.txt:5]: To install `scikit-learn` using pip we have to run ```pip install scikit-learn```. That is why I changed it in the requirements-test.txt (assuming that package will be installed during testing)

https://scikit-learn.org/0.16/install.html

To import `scikit-learn` the code is ```import sklearn```

File: api/.gitignore (hunk 15 above)
Comment (joysaikat) [reply api/.gitignore]: Done

File: requirements-test.txt (hunk 16 above)
Comment (itsjohnward) [reply requirements-test.txt:5]: `sklearn`'s an alias in PyPI for "latest version of `scikit-learn`" so version pinning doesn't work with it. Also, the alias does not exist in conda. Neither issue matters here in particular since this file is just for our unit tests, but in general `scikit-learn` is the better all-around name to use

Comment (joysaikat) [reply requirements-test.txt:5]: Right, I just tried ```pip install sklearn```, it seems to work. However, I clearly remember doing ```pip install sklearn==<>``` that did not work.

**Comment by nemozen** [comment]
almost there!  the notebook is still has a little bit of cruft that's not ideal for a show case. 

- can you remove the commented out code at the end?
- the table of contents is not that useful, and the links don't seem to work.. can you just remove it.
- the Report section can be improved.  either make it an actual code  block or remove it... instead of making it a code snippet in text to be pasted elsewhere.
- does logging.disable(level=logging.ERROR) override api.client.cfg.DEFAULT_LOG_LEVEL which is used in get_default_logger()? confusing it's better to just remove the local warnings manually

**Comment by joysaikat** [comment]
> almost there! the notebook is still has a little bit of cruft that's not ideal for a show case.
> 
> * can you remove the commented out code at the end?

Kept there as plotly produces cooler visualizations, but plotly plots could not be rendered in github. So, commented out if users wanted to download and use the notebook locally. Removed, in this version.

> * the table of contents is not that useful, and the links don't seem to work.. can you just remove it.

Discussed in slack. The links do not work because of github. They work fine if the notebook is downloaded locally or nbviewer is used. However, it makes the notebook longer. So, removed it.

> * the Report section can be improved.  either make it an actual code  block or remove it... instead of making it a code snippet in text to be pasted elsewhere.

Removed it.
> * does logging.disable(level=logging.ERROR) override api.client.cfg.DEFAULT_LOG_LEVEL which is used in get_default_logger()? confusing it's better to just remove the local warnings manually

Discussed in slack. I understand the need for a more concise notebook; hence, removed it. However, manually removing is a set back for reproducibility of the notebook.

**Comment by nemozen** [comment]
> > almost there! the notebook is still has a little bit of cruft that's not ideal for a show case.
> > 
> > * can you remove the commented out code at the end?
> 
> Kept there as plotly produces cooler visualizations, but plotly plots could not be rendered in github. So, commented out if users wanted to download and use the notebook locally. Removed, in this version.
> 
> > * the table of contents is not that useful, and the links don't seem to work.. can you just remove it.
> 
> Discussed in slack. The links do not work because of github. They work fine if the notebook is downloaded locally or nbviewer is used. However, it makes the notebook longer. So, removed it.
> 
> > * the Report section can be improved.  either make it an actual code  block or remove it... instead of making it a code snippet in text to be pasted elsewhere.
> 
> Removed it.
> 
> > * does logging.disable(level=logging.ERROR) override api.client.cfg.DEFAULT_LOG_LEVEL which is used in get_default_logger()? confusing it's better to just remove the local warnings manually
> 
> Discussed in slack. I understand the need for a more concise notebook; hence, removed it. However, manually removing is a set back for reproducibility of the notebook.

the user should be able to choose different log verbosity levels  when running it.  forcing the log level in code (with logging.disable) is bad practice because it would suppress other third party  warnings that the user may actually encounter (that may or may not occur in your case), even if the user wants to see warnings.   

more importantly, please don't forget to follow-up to address the underlying causes of the warnings. in the meantime, it's not helpful to have them pollute the notebook

**Review by nemozen — APPROVED** [review]
